query/src/sql/optimizer/memo.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+
+use common_exception::Result;
+
+use crate::sql::optimizer::group::Group;
+use crate::sql::optimizer::m_expr::MExpr;
+use crate::sql::optimizer::s_expr::SExpr;
+use crate::sql::IndexType;
+
+/// `Memo` is a search space which memoize possible plans of a query.
```

Comment:
Why do we name this struct `Memo`, are there any agreed-upon rules?

Reply:
This name comes from the paper of cascades optimizer. I think it can help people have knowledge about query optimizer get up with the project more easily.

Reply:
https://www.querifylabs.com/blog/memoization-in-cost-based-optimizers i guess this is related

Reply:
> https://www.querifylabs.com/blog/memoization-in-cost-based-optimizers i guess this is related

Yes, it's the state of memoization, just like a table in dynamic programming.

Reply:
We can have this refer URL inside the codes for other new guys to understand it in the future.

Reply:
> We can have this refer URL inside the codes for other new guys to understand it in the future.

I'm going to explain this in RFC, and refer the RFC in comment.